code/modules/security/genpop.dm

```diff
@@ -0,0 +1,650 @@
+				return FALSE
+			visible_message("<span class='warning'>[user] slots [I] into [src]'s ID slot, freeing its contents!</span>")
+			registered_name = null
+			desc = initial(desc)
+			locked = FALSE
+			update_icon()
+			qdel(I)
+			playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
+			return
+		else
+			to_chat(user, "<span class='danger'>Access Denied.</span>")
+	else
+		return ..()
+
+
+/obj/machinery/turnstile/Initialize(mapload)
+	. = ..()
+	icon_state = "turnstile"
+	//Attach a signal handler to the turf below for when someone passes through.
+	//Signal automatically gets unattached and reattached when we're moved.
+	var/static/list/loc_connections = list(
+		COMSIG_ATOM_ENTERED = PROC_REF(on_entered),
+	)
+	AddElement(/datum/element/connect_loc, loc_connections)
+
+/obj/machinery/turnstile/proc/on_entered(datum/source, atom/movable/arrived, atom/old_loc, list/atom/old_locs)
+	SIGNAL_HANDLER
+	if(istype(arrived, /mob))
+		flick("operate", src)
+		playsound(src,'sound/items/ratchet.ogg',50,0,3)
+
+///Handle movables (especially mobs) bumping into us.
+/obj/machinery/turnstile/Bumped(atom/movable/movable)
+	. = ..()
+	if(!istype(movable, /mob)) //just let non-mobs bump
+		return
+	if(machine_stat & BROKEN) //try to shock mobs if we're broken
+		try_shock(movable)
+		return
+	//pretend to be an airlock if a mob bumps us
+	//(which means they tried to move through but didn't have access)
+	flick("deny", src)
+	playsound(src,'sound/machines/deniedbeep.ogg',50,0,3)
+
+///Shock attacker if we're broken
+/obj/machinery/turnstile/attackby(obj/item/item, mob/user, params)
+	if(default_deconstruction_screwdriver(user, "turnstile", "turnstile", item))
+		update_appearance()//add proper panel icons
+		state = TURNSTILE_CIRCUIT_EXPOSED
+		return
+	if(item.tool_behaviour == TOOL_CROWBAR && panel_open && circuit != null)
+		to_chat(user, "<span class='notice'>You start tearing out the circuitry...")
+		if(do_after(user, 3 SECONDS))
+			circuit.forceMove(loc)
+			circuit = null
+			state = TURNSTILE_SHELL
+		return
+	. = ..()
+	if(machine_stat & BROKEN)
+		try_shock(user)
+
+//Shock attack if something is thrown at it if we're broken
+/obj/machinery/turnstile/hitby(atom/movable/AM, skipcatch, hitpush, blocked, datum/thrownthing/throwingdatum)
+	if(isobj(AM))
+		if(prob(50) && BROKEN)
+			var/obj/O = AM
+			if(O.throwforce != 0)//don't want to let people spam tesla bolts, this way it will break after time
+				playsound(src, 'sound/magic/lightningshock.ogg', 100, 1, extrarange = 5)
+				tesla_zap(src, 3, 8000, TESLA_MOB_DAMAGE | TESLA_OBJ_DAMAGE | TESLA_MOB_STUN | TESLA_ALLOW_DUPLICATES) // Around 20 damage for humans
+	return ..()
+
+/obj/machinery/turnstile/welder_act(mob/living/user, obj/item/I)
+	//Shamelessly copied airlock code
+	. = TRUE
+	if(circuit == null && user.a_intent == INTENT_HARM)
+		var/obj/item/weldingtool/W = I
+		if(W.welding)
```

Comment:
Calling welder_act doesn't imply that the item is a welder, for example plasma cutters are actually welders too. Instead of doing W.welding, you should use the generic I.tool_start_check instead which for welders ensures that the welder is capable of welding. Note that use_tool also performs these checks, so you shouldn't actually need them unless you are sleeping; in this case you can just remove W.welding check.